src/Symfony/Component/Validator/Constraints/File.php

```diff
@@ -58,29 +57,57 @@ class File extends Constraint
     public $uploadExtensionErrorMessage = 'A PHP extension caused the upload to fail.';
     public $uploadErrorMessage = 'The file could not be uploaded.';
 
+    protected $maxSize;
+
     public function __construct($options = null)
     {
         parent::__construct($options);
 
         if ($this->maxSize) {
-            if (ctype_digit((string) $this->maxSize)) {
-                $this->maxSize = (int) $this->maxSize;
-                $this->binaryFormat = null === $this->binaryFormat ? false : $this->binaryFormat;
-            } elseif (preg_match('/^\d++k$/i', $this->maxSize)) {
-                $this->maxSize = $this->maxSize * 1000;
-                $this->binaryFormat = null === $this->binaryFormat ? false : $this->binaryFormat;
-            } elseif (preg_match('/^\d++M$/i', $this->maxSize)) {
-                $this->maxSize = $this->maxSize * 1000000;
-                $this->binaryFormat = null === $this->binaryFormat ? false : $this->binaryFormat;
-            } elseif (preg_match('/^\d++Ki$/i', $this->maxSize)) {
-                $this->maxSize = $this->maxSize << 10;
-                $this->binaryFormat = null === $this->binaryFormat ? true : $this->binaryFormat;
-            } elseif (preg_match('/^\d++Mi$/i', $this->maxSize)) {
-                $this->maxSize = $this->maxSize << 20;
-                $this->binaryFormat = null === $this->binaryFormat ? true : $this->binaryFormat;
-            } else {
-                throw new ConstraintDefinitionException(sprintf('"%s" is not a valid maximum size', $this->maxSize));
-            }
+            $this->evaluateBinaryFormat();
+        }
+    }
+
+    public function __set($option, $value)
+    {
+        if ('maxSize' === $option) {
+            $this->maxSize = $value;
+            $this->evaluateBinaryFormat();
```

Comment:
the strange thing here is that you first set the value, then validate it, which means that by catching the exception, you can force an invalid value in maxSize. Is it intended?